libgalois/include/katana/ResultReduction.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <optional>
+
+#include "katana/Result.h"
+
+namespace katana {
+
+namespace internal {
+
+// TODO(roshan) combine/merge two error infos
+struct AnyCopyableErrorInfo {
+  std::optional<CopyableErrorInfo> operator()(
+      const std::optional<CopyableErrorInfo>& lhs,
+      const std::optional<CopyableErrorInfo>& rhs) const {
+    if (lhs) {
+      return *lhs;
+    }
+    if (rhs) {
+      return *rhs;
+    }
+    return std::nullopt;
+  }
+};
+
+struct IdentityCopyableErrorInfo {
+  std::optional<CopyableErrorInfo> operator()() const { return std::nullopt; }
+};
+
+}  // namespace internal
+
+class GReduceCopyableErrorInfo
+    : public Reducible<
+          std::optional<CopyableErrorInfo>, internal::AnyCopyableErrorInfo,
+          internal::IdentityCopyableErrorInfo> {
+  using base_type = Reducible<
+      std::optional<CopyableErrorInfo>, internal::AnyCopyableErrorInfo,
+      internal::IdentityCopyableErrorInfo>;
+
+public:
+  GReduceCopyableErrorInfo()
+      : base_type(
+            internal::AnyCopyableErrorInfo(),
+            internal::IdentityCopyableErrorInfo()) {}
+};
+
+#define KATANA_REDUCTION_CHECKED_IMPL(                                         \
+    error_info, result_name, expression, ...)                                  \
+  ({                                                                           \
+    auto result_name = (expression);                                           \
+    if (::katana::internal::CheckedExpressionFailed(result_name)) {            \
+      error_info.update(                                                       \
+          ::katana::internal::CheckedExpressionToError(result_name)            \
+              .WithContext(__VA_ARGS__)                                        \
+              .WithContext(FMT_STRING("({}:{})"), __FILE__, __LINE__));        \
+      return;                                                                  \
+    }                                                                          \
+    std::move(                                                                 \
+        ::katana::internal::CheckedExpressionToValue(std::move(result_name))); \
+  })
+
+/// KATANA_REDUCTION_CHECKED_CONTEXT is similar to KATANA_CHECKED_CONTEXT
+/// except that instead of returning the error, it copies the error
+/// into error_info by calling error_info.update(<CopyableErrorInfo>)
```

Comment:
If you write `GReduceCopyableErrorInfo::update(CopyableErrorInfo)` here instead it will turn into a link in the documentation. Might be nicer that way.